okio/src/main/java/okio/TeeSource.java

```diff
@@ -0,0 +1,41 @@
+package okio;
+
+import java.io.IOException;
+
+
+class TeeSource implements Source {
+
+  private final Source source;
+  private final BufferedSink cacheBody;
+
+  public TeeSource(Source source, BufferedSink cacheBody) {
+    this.source = source;
+    this.cacheBody = cacheBody;
+  }
+
+  @Override
+  public long read(Buffer sink, long byteCount) throws IOException {
+    long bytesRead;
+    bytesRead = source.read(sink, byteCount);
+
+    if (bytesRead == -1) {
+      cacheBody.close(); // The cache response is complete!
```

Comment:
Not sure we want to retain this close here from the cache code for the general case.